### PR TITLE
#35739-the-questions-system

### PR DESCRIPTION
**Pull request by alexluca06**

### CODE REVIEW

File: lib/redmine_goodies_macros.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
+module RedmineGoodies
+    module Macros
+        # Retrieves a mapping of note references (that contain questions using the question macro) to their corresponding answers.
+        def self.get_all_answers(issue_id)
+            issue = Issue.find_by(id: issue_id)
```
Comment (mircea-negreanu) [inline lib/redmine_goodies_macros.rb]: I don't like that you need to make at least 2 queries to the database (one gets the issue, the other gets the journals ordered).

Isn't something disponible when the macro is run? E.g. `@journal`, `@...`?

The problem is that those queries will be done each time the macro is rendered.

File: lib/redmine_goodies_macros.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        def self.get_all_answers(issue_id)
+            issue = Issue.find_by(id: issue_id)    
+            journals = issue.journals.order(:created_on)
+
+            # eg. {"#note-19"=>[{:question=>"1", :response_from=>"#note-20"}, {:question=>"2", :response_from=>"#note-20"}
+            answered_questions = {}
+
+            # eg. "{{answer(#note-56, 1/}}", the regex below will extract:
+            # '#note-56' and '1'
+            regex = /\{\{answer\((#\w+-\d+),\s*(\d+)\/\)\}\}/
```
Comment (mircea-negreanu) [inline lib/redmine_goodies_macros.rb]: What happens if we have `{` inside the answer?

File: lib/redmine_goodies_macros.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                        answered_questions[note_ref] << { question: question_number, response_from: response_from_note_id }
+                    end
+                end
+            end
+            answered_questions
+        end
+
+        Redmine::WikiFormatting::Macros.register do
+            desc "Custom macro for rendering questions"
+            macro :questions do |obj, args, text|
```
Comment (mircea-negreanu) [inline lib/redmine_goodies_macros.rb]: I'm wondering if it is possible to restrict a macro to only issues. I think it is also available on wiki, correct?

File: lib/redmine_goodies_macros.rb (hunk 1 above)
Comment (mircea-negreanu) [reply lib/redmine_goodies_macros.rb]: If you don't have that information, maybe get only the notes that have the answer macro in them. I'm wondering about the performance implication of a task with too many notes (e.g. > 1000)

Comment (cristian-spiescu) [reply lib/redmine_goodies_macros.rb]: I don't want this macro to generate impact in the DB!

So the cross linking should be done via JS, client side.

File: lib/redmine_goodies_macros.rb (hunk 3 above)
Comment (alexluca06) [reply lib/redmine_goodies_macros.rb]: Yes, they are also available on the wiki, but I added `if obj.is_a?(Journal)`, which restricts the usage to Issues only.

File: lib/redmine_goodies_macros.rb (hunk 2 above)
Comment (alexluca06) [reply lib/redmine_goodies_macros.rb]: 1.  {{answer(#note-50, 1/)}} {example} -> [Answer for [#note-50](http://localhost:4000/issues/3#note-50), 1/] {example}

2. {{answer( { #note-50 }, 1/)}} example -> Using '{' inside answer macro is not allowed. The answer macro is added automatically in the editor, and the user should not edit it.

File: lib/redmine_goodies_macros.rb (hunk 1 above)
Comment (alexluca06) [reply lib/redmine_goodies_macros.rb]: I moved the cross linking on client side (javascript)

**Pull request merged**